### DISCUSSION TITLE
HOGWARTS: Create Unique IDs to Delete Elements from Array using method chaining and ternary operator

### DISCUSSION THREAD

**Discussion by danicrosby** [nss-evening-cohort-14/evening-client-side#65] [Help Me!] [answered]
Your Name: Dani Crosby

Have you searched the other help tickets? Yes

Describe your issue: (See video) I am able to get my card(s) to delete, but only the first one. I believe this is an id/event target issue, but I don't know how to get past it! I've searched all over the nets, re-watched the videos including the delete parts

List any error messages that appear : No errors appear

What have you googled? 
I've tried to increment the card id globally,
https://stackoverflow.com/questions/47962992/how-increment-id-object-with-js/47963031
https://dzone.com/articles/removing-element-plain

GitHub File: https://github.com/danicrosby/sorting-hat/blob/sort-btn/main.js

Give me a comment below letting us know that you understand these items or let us know in the comments if you have any questions.

10-4

LOOM:
https://www.loom.com/share/d7f8e8c7702f4cf8abb59fafd65824c2

**Comment by drteresavasquez** [comment]
One of the issues is that the remove method you're using in delete is not removing the element from your data and ONLY from the DOM. Go back and look at how we handled delete in pies.

**Reply by drteresavasquez** [reply]
Another issue is that you are not using a unique ID on either the card or the button. How will the DOM or function know what to delete?

**Comment by drteresavasquez** [comment]
@danicrosby I see you gave a thumbs up on one of the responses. Were you able to get to a solution? If so, can you tell us what worked? If not, let us see your updated code so we can continue to help you through the issue.

**Reply by danicrosby** [reply]
no ;( I spent the last few hours 'refactoring' and watching the videos. I'm in a downward spiral and keep coming back to my original code to see if there's a logical fix. I'm happy to write it all over again, but I'm confused on the PIES example as both the array is named PIES and the container it lives in on the index is also called PIES. I'm currently working on how to implement the for.Each method, would be SUPER interested to see if I'm like lukewarm with the code I already have. HELP! Hahah

**Reply by drteresavasquez** [reply]
Lemme check it out. Is it at the same link?

**Reply by danicrosby** [reply]
yeah-you can check the same file out-  I've made some ugly changes that I'm not proud of  LOL

**Reply by danicrosby** [reply]
actually- I went ahead and push the ugly code I was trying to refactor- same results(ish)

**Reply by danicrosby** [reply]
nevermind. My brain hurts, just look at the original file!!!! Thanks in advance

**Reply by drteresavasquez** [reply]
So, in the object here: https://github.com/danicrosby/sorting-hat/blob/sort-btn/main.js#L63-L66 you have...

```
const obj = {
    name,
    house,
  };
```

## Add an ID to your student Object
You may consider having an ID inside of the object when it is created. You could probably get it to assign a random ID or use something to create that like this maybe:

```
// get all the student IDs and create a new array and sort them lowest to highest
const studentIds = students.map(student => student.id).sort((a, b) => a - b);

// ternary operator: if the students array is not empty, create an ID that is +1 of the last item in the array. Otherwise make the id 1
const id = studentIds.length ? studentIds[(studentIds.length - 1)] + 1 : 1;

const obj = {
    name,
    house,
    id,
  };
```

## Add the ID of the student to your cards
Then, inside of your card creator, you can add that ID to the card and as an ID on the button, so when your click event fires and you get the `e.target.id` AND the `e.target.type === 'button'` then we can go into our students array and find the index of that item using a `.findIndex()` like so:

## Add this to your delete so that the original array is modified
```
// this method returns the index of the object you are trying to remove in the array
const studentIndex = students.findIndex(student => student.id === 1);

// this removes the item from the array - NOTE: this modifies the original array
students.splice(studentIndex, 1)
```
## Put is all together by calling the students array again
After you do that, you can call your card builder and pass it the `students` array again and your card should be removed from the DOM and your array is modified.

### So to recap:
1. Add an ID to the objects you create, so this means that initially, your students array should be empty
2. When a user creates a student, add the ID using the sample code or something similar. This will ensure that the IDs are unique.
3. Ensure that your data is updated when you 'expel' a student.
4. In the delete, call the card builder function again and pass it the modified students array

NOTE: In order to get to the stretch goals, one of them is to have expelled students in another div on the DOM, so you would modify the delete function to grab that object and add it to another array. I would use the spread operator to add it. We can get to that in another ticket.

**Comment by drteresavasquez** [comment, marked as answer]
Posted outside of sub comments for view:

So, in the object here: https://github.com/danicrosby/sorting-hat/blob/sort-btn/main.js#L63-L66 you have...

```
const obj = {
    name,
    house,
  };
```

## Add an ID to your student Object
You may consider having an ID inside of the object when it is created. You could probably get it to assign a random ID or use something to create that like this maybe:

```
// get all the student IDs and create a new array and sort them lowest to highest
const studentIds = students.map(student => student.id).sort((a, b) => a - b);

// ternary operator: if the students array is not empty, create an ID that is +1 of the last item in the array. Otherwise make the id 1
const id = studentIds.length ? studentIds[(studentIds.length - 1)] + 1 : 1;

const obj = {
    name,
    house,
    id,
  };
```

## Add the ID of the student to your cards
Then, inside of your card creator, you can add that ID to the card and as an ID on the button, so when your click event fires and you get the `e.target.id` AND the `e.target.type === 'button'` then we can go into our students array and find the index of that item using a `.findIndex()` like so:

## Add this to your delete so that the original array is modified
```
// this method returns the index of the object you are trying to remove in the array
const studentIndex = students.findIndex(student => student.id === 1);

// this removes the item from the array - NOTE: this modifies the original array
students.splice(studentIndex, 1)
```
## Put is all together by calling the students array again
After you do that, you can call your card builder and pass it the `students` array again and your card should be removed from the DOM and your array is modified.

### So to recap:
1. Add an ID to the objects you create, so this means that initially, your students array should be empty
2. When a user creates a student, add the ID using the sample code or something similar. This will ensure that the IDs are unique.
3. Ensure that your data is updated when you 'expel' a student.
4. In the delete, call the card builder function again and pass it the modified students array

NOTE: In order to get to the stretch goals, one of them is to have expelled students in another div on the DOM, so you would modify the delete function to grab that object and add it to another array. I would use the spread operator to add it. Here is a ticket discussing that #72 

## Links to suggestions above: 
1. Ternary Operator: https://www.javascripttutorial.net/javascript-ternary-operator/
2. Example of deleting pies: https://github.com/nss-evening-cohort-14/event-listeners/blob/delete-pies/main.js#L168-L178

**Reply by danicrosby** [reply]
Ok! Thanks!! I'll make the appropriate changes and check back in. Appreciate it.

**Reply by drteresavasquez** [reply]
Of course! I want to walk through this in class tonight. You mind being the example? 😃

**Reply by danicrosby** [reply]
Yes please!!! :)

**Comment by danicrosby** [comment]
A long winded overview of our issue ticket follow up! Thanks for much for the help everyone!

https://www.loom.com/share/ecaa1f186f9c4c76af49d94cf00d3191

**Reply by drteresavasquez** [reply]
Love the video. There are some things that you may need help with clarifying, but that is all in the process of learning. I think you did an awesome job explaining and this is something that you can review as the answer above writes it out for you! Thank you! 🥳
